pkg/genericapiserver/storage_factory.go

```diff
@@ -255,16 +254,7 @@ func NewStorageCodec(storageMediaType string, ns runtime.StorageSerializer, stor
 	}
 
 	ds := recognizer.NewDecoder(s, ns.UniversalDeserializer())
-	encoder := ns.EncoderForVersion(s, storageVersion)
-	decoder := ns.DecoderToVersion(ds, memoryVersion)
-	if memoryVersion.Group != storageVersion.Group {
-		// Allow this codec to translate between groups.
-		if err := versioning.EnableCrossGroupEncoding(encoder, memoryVersion.Group, storageVersion.Group); err != nil {
-			return nil, fmt.Errorf("error setting up encoder from %v to %v: %v", memoryVersion, storageVersion, err)
-		}
-		if err := versioning.EnableCrossGroupDecoding(decoder, storageVersion.Group, memoryVersion.Group); err != nil {
-			return nil, fmt.Errorf("error setting up decoder from %v to %v: %v", storageVersion, memoryVersion, err)
-		}
-	}
+	encoder := ns.EncoderForVersion(s, runtime.NewMultiGroupVersioner(storageVersion, storageVersion.Group, memoryVersion.Group))
```

Comment:
Yeah, it is probably better to explicitly declare these as encoders for a
group kind.

On Mon, Jun 6, 2016 at 8:20 AM, David Eads <notifications@github.com> wrote:

> In pkg/genericapiserver/storage_factory.go
> <https://github.com/kubernetes/kubernetes/pull/26044#discussion_r65880328>
> :
>
> > @@ -255,16 +254,7 @@ func NewStorageCodec(storageMediaType string, ns runtime.StorageSerializer, stor
> >  	}
> >
> >  	ds := recognizer.NewDecoder(s, ns.UniversalDeserializer())
> > -	encoder := ns.EncoderForVersion(s, storageVersion)
> > -	decoder := ns.DecoderToVersion(ds, memoryVersion)
> > -	if memoryVersion.Group != storageVersion.Group {
> > -		// Allow this codec to translate between groups.
> > -		if err := versioning.EnableCrossGroupEncoding(encoder, memoryVersion.Group, storageVersion.Group); err != nil {
> > -			return nil, fmt.Errorf("error setting up encoder from %v to %v: %v", memoryVersion, storageVersion, err)
> > -		}
> > -		if err := versioning.EnableCrossGroupDecoding(decoder, storageVersion.Group, memoryVersion.Group); err != nil {
> > -			return nil, fmt.Errorf("error setting up decoder from %v to %v: %v", storageVersion, memoryVersion, err)
> > -		}
> > -	}
> > +	encoder := ns.EncoderForVersion(s, runtime.NewMultiGroupVersioner(storageVersion, storageVersion.Group, memoryVersion.Group))
>
> Should we start forcing GroupKind to be passed in addition to
> GroupResource in the method calls above? Not necessarily in this pull,
> but I think it may make the flow a little more obvious.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26044/files/a16b294e553924a3213ad4119f03c3ba2953965e#r65880328>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p5oIoFMLe83jl82P75ngfvVPETekks5qJBCGgaJpZM4IkFHI>
> .
>